Exapnd MachineConfigPool for creation

yossisegev opened the pull request:
Support creation of MachineConfigPool resource, with several attributes.

myakove commented on ocp_resources/machine_config_pool.py:
```diff
@@ -7,3 +7,30 @@ class MachineConfigPool(Resource):
         UPDATED = "Updated"
         UPDATING = "Updating"
+
+    def __init__(
```
Please check other resources, all should be with default None and you have missing args that should be here.

yossisegev replied:
@myakove Which args are missing? Can you please provide a resource which I can use as a reference to complete the missing args?
And BTW - why all args should be with default None? Doesn't it make sense that some of them must be set by the caller?

myakove replied:
The user have option to send a YAML file, therefor no args needed.

yossisegev replied:
Ack, thanks.

myakove replied:
https://github.com/RedHatQE/openshift-python-wrapper/blob/main/CONTRIBUTING.md

yossisegev replied:
Thanks Meni, now I added the extra missing args.

yossisegev replied:
Ack

myakove commented on ocp_resources/machine_config_pool.py:
```diff
@@ -7,3 +7,30 @@ class MachineConfigPool(Resource):
+        self.machine_config_selector_matches = machine_config_selector_matches
+
+    def to_dict(self):
+        super().to_dict()
```
Check how this is done in other resources, need to handle yaml_file

yossisegev replied:
Why must it handle yaml_file if I'm not going to send a yaml file to this class?

myakove replied:
Maybe you are not going to send any YAML, but this resource is for everyone to use not just for your use case.

myakove replied:
I don't see where this was resolved

yossisegev replied:
Sorry, I didn't notice I didn't handle yaml_file here in to_dict().

myakove commented on ocp_resources/machine_config_pool.py:
```diff
@@ -7,3 +7,30 @@ class MachineConfigPool(Resource):
+    def to_dict(self):
+        super().to_dict()
+        if self.resource_labels:
+            self.res["metadata"] = []
```
can be done in one line with list comprehension

yossisegev replied:
Done

myakove commented on ocp_resources/machine_config_pool.py:
```diff
@@ -7,3 +7,30 @@ class MachineConfigPool(Resource):
+            for label in self.resource_labels:
+                self.res["metadata"].append(label)
+
+        self.res["spec"]["configuration"] = (
```
`if self.configuration:` ....

yossisegev replied:
Done

myakove commented on ocp_resources/machine_config_pool.py:
```diff
@@ -7,3 +7,30 @@ class MachineConfigPool(Resource):
+            self.configuration if self.configuration else {}
+        )
+
+        self.res["spec"]["machineConfigSelector"]["matchExpressions"] = []
```
ditto, use list comprehension

yossisegev replied:
Done

rnetser reviewed (changes requested):
A reference to the api is missing; we cannot review it without it. We need to see which fields are requierd

rnetser commented on ocp_resources/machine_config_pool.py:
```diff
@@ -2,8 +2,60 @@
+    def to_dict(self):
+        super().to_dict()
+        if not self.yaml_file:
+            if self.resource_labels:
+                self.res["metadata"]["labels"] = self.resource_labels
```
this already exists in resource

yossisegev replied:
Ack for both

rnetser commented on ocp_resources/machine_config_pool.py:
```diff
@@ -2,8 +2,60 @@
+        max_unavailable=None,
+        paused=none,
+    ):
+        super().__init__(name=name, yaml_file=yaml_file)
```
missing docstring

yossisegev replied:
Ack (kwargs and docstring).

rnetser commented on ocp_resources/machine_config_pool.py:
```diff
@@ -2,8 +2,60 @@
+
+            self.res["spec"] = {}
+            self.res["spec"]["configuration"] = {}
+            if self.configuration:
```
`configuration` is mandatory ; please raise if not passed

yossisegev replied:
Actually, it's not mandatory.
You can apply the MCP manifest without the `configuration` specified, and it will be generated automatically.

rnetser commented on ocp_resources/machine_config_pool.py:
```diff
@@ -2,8 +2,60 @@
+        self,
+        name=None,
+        labels=None,
+        machine_config_selector_matches=None,
```
```suggestion
        machine_config_selector=None,
```
use the attribute names as much as possible; the docstring should describe what this attribute is for

yossisegev replied:
Ack

rnetser commented on ocp_resources/machine_config_pool.py:
```diff
@@ -2,8 +2,60 @@
+        machine_config_selector_matches=None,
+        configuration=None,
+        yaml_file=None,
+        node_selector_labels=None,
```
ditto

yossisegev replied:
Ack

rnetser commented on ocp_resources/machine_config_pool.py:
```diff
@@ -2,8 +2,60 @@
+        yaml_file=None,
+        node_selector_labels=None,
+        max_unavailable=None,
+        paused=none,
```
should be `None`

yossisegev replied:
Ack

rnetser commented on ocp_resources/machine_config_pool.py:
```diff
@@ -2,8 +2,60 @@
+            self.res["spec"]["configuration"] = {}
+            if self.configuration:
+                self.res["spec"]["configuration"] = self.configuration
+
```
if self.machine_config_selector...
unless this is mandatory and then you should check before assigning if this value was passed

yossisegev replied:
Ack

rnetser commented on ocp_resources/machine_config_pool.py:
```diff
@@ -2,8 +2,69 @@
+    def __init__(
+        self,
+        name=None,
+        labels=None,
```
`labels` should be removed; will be part of `kwargs`

yossisegev replied:
Ack

rnetser commented on ocp_resources/machine_config_pool.py:
```diff
@@ -2,8 +2,69 @@
+        paused=None,
+        **kwargs,
+    ):
+        super().__init__(name=name, yaml_file=yaml_file, label=labels, **kwargs)
```
```suggestion
        super().__init__(name=name, yaml_file=yaml_file, **kwargs)
```

rnetser commented on ocp_resources/machine_config_pool.py:
```diff
@@ -2,8 +2,69 @@
+    def to_dict(self):
+        super().to_dict()
+        if not self.yaml_file:
+            self.res["spec"] = {}
+            self.res["spec"]["configuration"] = {}
+            if self.configuration:
+                self.res["spec"]["configuration"] = self.configuration
```
```suggestion
            self.res.update(
                {
                    "spec": {
                        "configuration": self.configuration or {},
                    }
                }
            )
```

yossisegev replied:
Ack

rnetser reviewed (changes requested):
Some comments are not addressed

yossisegev commented:
From what I see, I addressed all comments.
I left some comments unresolved, in these comments I responded to review comments I got.
From my POV I can mark these comments as resolved, if you accept my responses.
If I missed any comment (or if my response is not acceptable) - please let me know which comment it is.

myakove commented on ocp_resources/machine_config_pool.py:
```diff
@@ -2,8 +2,72 @@
+    MachineConfigPool object. API reference:
+    https://docs.openshift.com/container-platform/4.12/rest_api/machine_apis/machineconfigpool-machineconfiguration-openshift-io-v1.html
+
+    Args:
```
Let's make is simple, please remove all things from docstring leave only what your resource needs.
in `__init__` get only what your resource needs and then **kwargs

yossisegev replied:
Ack.
I think I got it now.

rnetser commented on ocp_resources/machine_config_pool.py:
```diff
@@ -2,8 +2,70 @@
+
+    Args:
+        name (string): Object name.
+        machine_config_selector (object): List of matching labels, to determine on which nodes to apply this
```
machineConfigSelector -> matchLabels get a dict of label_name; label_value
```
  machineConfigSelector:
    matchLabels:
      machineconfiguration.openshift.io/role: master
```
matchExpressions gets a list of dicts

yossisegev replied:
Ack

rnetser commented on ocp_resources/machine_config_pool.py:
```diff
@@ -2,8 +2,70 @@
+    https://docs.openshift.com/container-platform/4.12/rest_api/machine_apis/machineconfigpool-machineconfiguration-openshift-io-v1.html
+
+    Args:
+        name (string): Object name.
```
can be removed; only resource-specific args should be set along with `kwargs`

yossisegev replied:
Ack

rnetser commented on ocp_resources/machine_config_pool.py:
```diff
@@ -2,8 +2,70 @@
+        machine_config_selector (object): List of matching labels, to determine on which nodes to apply this
+            MachineConfigPool object.
+        configuration (object): Current MachineConfig object for the machine config pool.
+        yaml_file (str): yaml file for the resource.
```
can be removed (part of `wargs`)

yossisegev replied:
Ack

rnetser commented on ocp_resources/machine_config_pool.py:
```diff
@@ -2,8 +2,70 @@
+        name (string): Object name.
+        machine_config_selector (object): List of matching labels, to determine on which nodes to apply this
+            MachineConfigPool object.
+        configuration (object): Current MachineConfig object for the machine config pool.
```
```suggestion
        configuration (dict): {"name": <str; add exaplanation>, "source": <list; add exaplanation> and an example
```

yossisegev replied:
Ack

rnetser commented on ocp_resources/machine_config_pool.py:
```diff
@@ -2,8 +2,70 @@
+            if self.node_selector_labels:
+                self.res["spec"]["nodeSelector"] = {}
+                self.res["spec"]["nodeSelector"][
+                    "matchLabels"
```
can also be other things, for example `matchExpressions`  https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#resources-that-support-set-based-requirements

yossisegev replied:
Moved to kwargs, as it is already handled in `Resource`.

yossisegev replied:
Ack

rnetser commented on ocp_resources/machine_config_pool.py:
```diff
@@ -2,8 +2,70 @@
+            MachineConfigPool object.
+        configuration (object): Current MachineConfig object for the machine config pool.
+        yaml_file (str): yaml file for the resource.
+        node_selector_labels (object (string)): Map of {key,value} pairs that determine node selection criteria.
```
same logic as `machineConfigSelector`  applies here

yossisegev replied:
Ack

rnetser commented on ocp_resources/machine_config_pool.py:
```diff
@@ -2,8 +2,70 @@
+            if self.machine_config_selector:
+                self.res["spec"]["machineConfigSelector"] = {}
+                self.res["spec"]["machineConfigSelector"][
+                    "matchExpressions"
```
see comment below on options for this selector

yossisegev replied:
Ack

rnetser commented on ocp_resources/machine_config_pool.py:
```diff
@@ -2,8 +2,70 @@
+        configuration (object): Current MachineConfig object for the machine config pool.
+        yaml_file (str): yaml file for the resource.
+        node_selector_labels (object (string)): Map of {key,value} pairs that determine node selection criteria.
+        max_unavailable (integer/string): Number/percentage of nodes that can go Unavailable during an update.
```
```suggestion
        max_unavailable (int/str): Number/percentage of nodes that can go Unavailable during an update.
```

yossisegev replied:
Ack

rnetser commented on ocp_resources/machine_config_pool.py:
```diff
@@ -2,8 +2,70 @@
+        yaml_file (str): yaml file for the resource.
+        node_selector_labels (object (string)): Map of {key,value} pairs that determine node selection criteria.
+        max_unavailable (integer/string): Number/percentage of nodes that can go Unavailable during an update.
+        paused (object (boolean)): Whether changes to this MachineConfigPool should be stopped.
```
```suggestion
        paused (bool): Whether changes to this MachineConfigPool should be stopped.
```

yossisegev replied:
Ack

myakove commented on ocp_resources/machine_config_pool.py:
```diff
@@ -2,8 +2,74 @@
+    https://docs.openshift.com/container-platform/4.12/rest_api/machine_apis/machineconfigpool-machineconfiguration-openshift-io-v1.html
+
+    Args:
+        machine_config_selector (object): Matching labels/expressions, to determine which MachineConfig objects
```
What do you mean by `(object)` ? what should the user send here?

yossisegev replied:
The user should send either a `matchExpressions` or a `matchLabels`  object that will be put in the `machineConfigSelector` object, as it is specified in the API. That's why I wrote "Matching labels/expressions".

myakove replied:
Python object? do have such one?

yossisegev replied:
Actually it's a dict.
I changed the type.

myakove commented on ocp_resources/machine_config_pool.py:
```diff
@@ -2,8 +2,74 @@
+            MachineConfigPool object.
+        node_selector_expressions (object): Matching expressions, to determine to which nodes to apply this
+            MachineConfigPool object.
+        max_unavailable (int/str): Number/percentage of nodes that can go Unavailable during an update.
```
`(int/str)` should be int.

rnetser replied:
based on api, could be both, please replace with `int or str`

yossisegev replied:
Not accroding to the [API](https://docs.openshift.com/container-platform/4.12/rest_api/machine_apis/machineconfigpool-machineconfiguration-openshift-io-v1.html#spec)

myakove commented on ocp_resources/machine_config_pool.py:
```diff
@@ -2,8 +2,74 @@
+        node_selector_expressions (object): Matching expressions, to determine to which nodes to apply this
+            MachineConfigPool object.
+        max_unavailable (int/str): Number/percentage of nodes that can go Unavailable during an update.
+        paused (bool): Whether changes to this MachineConfigPool should be stopped.
```
`should be paused.`

yossisegev replied:
The description of this member in the API says "should be stopped". Do you still want me to change it?

myakove replied:
leave it as the API say

rnetser commented on ocp_resources/machine_config_pool.py:
```diff
@@ -2,8 +2,74 @@
+            to apply this MachineConfigPool object.
+        configuration (dict): Targeted MachineConfig object for the machine config pool, in the following format:
+            {"name": (str), "source": <List of dicts, each representing a MachineConfig resource>}
+        node_selector_labels (object): Matching labels, to determine to which nodes to apply this
```
```suggestion
        node_selector (dict): nodeSelector dict with supported selector logic
```
For example:


  matchLabels:
    component: redis
  matchExpressions:
    - { key: tier, operator: In, values: [cache] }
    - { key: environment, operator: NotIn, values: [dev] }

yossisegev replied:
Ack

rnetser replied:
This is not resolved

rnetser commented on ocp_resources/machine_config_pool.py:
```diff
@@ -2,8 +2,74 @@
+            {"name": (str), "source": <List of dicts, each representing a MachineConfig resource>}
+        node_selector_labels (object): Matching labels, to determine to which nodes to apply this
+            MachineConfigPool object.
+        node_selector_expressions (object): Matching expressions, to determine to which nodes to apply this
```
should be `machine_config_selector`
same logic as `node_selector`

yossisegev replied:
Sorry, I don't understand this comment.
What do you mean by "should be machine_config_selector"?
These are members of nodeSelector.

rnetser replied:
`MachineConfigPool` has both `machineConfigSelector` and `nodeSelector`:
```
  machineConfigSelector:
    matchLabels:
      machineconfiguration.openshift.io/role: master
  nodeSelector:
    matchLabels:
      node-role.kubernetes.io/master: ""
```

Each can have different options for selection logic.
Selection logic has `matchLabels` and other options as well

yossisegev replied:
I hope I got what you mean this time.

yossisegev commented:
/verified

yossisegev reviewed (commented):
/verified

yossisegev commented on ocp_resources/machine_config_pool.py:
```diff
@@ -2,8 +2,89 @@
+            if self.machine_config_selector:
+                self.res["spec"]["machineConfigSelector"] = self.machine_config_selector
+
+            if self.node_selector_labels or self.node_selector_expressions:
+                self.res["spec"]["nodeSelector"] = {}
+                if self.node_selector_labels:
+                    self.res["spec"]["nodeSelector"][
+                        "matchLabels"
+                    ] = self.node_selector_labels
+                if self.node_selector_expressions:
+                    self.res["spec"]["nodeSelector"][
+                        "matchExpressions"
+                    ] = self.node_selector_expressions
+
```
But `Resource` class doesn't handle `node_selector` other than the standard hostname nodeSelector.
MCP has nodeSelector of different types - `matchLabels` and `matchExpressions`, therefore I added this special handling.

yossisegev replied:
Done

rnetser commented on ocp_resources/machine_config_pool.py:
```diff
@@ -2,8 +2,78 @@
+            )
+
+            if self.machine_config_selector:
+                self.res["spec"]["machineConfigSelector"] = self.machine_config_selector
```
`self.res["spec"]` - save and re-use

yossisegev replied:
Done

rnetser commented on ocp_resources/machine_config_pool.py:
```diff
@@ -2,8 +2,78 @@
+    https://docs.openshift.com/container-platform/4.12/rest_api/machine_apis/machineconfigpool-machineconfiguration-openshift-io-v1.html
+
+    Args:
+        machine_config_selector (dict): Matching labels/expressions, to determine which MachineConfig objects
+            to apply this MachineConfigPool object.
+            For filtering based on labels, the `matchLabels` dict is used - the same way as it is used in the
+                nodeSelector (see the example of node_selector["matchLabels"] below).
+            For filtering based on expressions, the `matchExpressions` dict is used - the same way as it is used in the
+                nodeSelector (see the example of node_selector["matchExpressions"] below).
```
```suggestion
        machine_config_selector (dict): Matching labels/expressions. <add example here>
```

yossisegev replied:
Instead of duplicating the same example, I added an explicit reference to the examples that I added for the nodeSelector members.

rnetser commented on ocp_resources/machine_config_pool.py:
```diff
@@ -2,8 +2,78 @@
+                nodeSelector (see the example of node_selector["matchExpressions"] below).
+        configuration (dict): Targeted MachineConfig object for the machine config pool, in the following format:
+            {"name": (str), "source": <List of dicts, each representing a MachineConfig resource>}
+        node_selector (dict): Matching dict with supported selector logic, either labels or expressions.
+            matchLabels example:
+            matchLabels:
+                component: <some component>
+                matchExpressions:
+                - { key: tier, operator: In, values: [cache] }
+                - { key: environment, operator: NotIn, values: [dev] }
+            matchExpressions example:
+            matchExpressions:
+            - key: <resource name>/role
+              operator: In
+              values:
+              - value_1
+              - value_2
```
```suggestion
        node_selector (dict): Matching dict with supported selector logic, either labels or expressions. See example under `machine_config_selector`
```

rnetser replied:
Because when a user reads the docstring they should have all the info in one place. One param (preferably the first one in the docstring) should hold the explanation and example. The other can refer to it.

yossisegev commented:
Yeah, whatever.
Done (hopefully).

yossisegev commented:
@rnetser @myakove A kind reminder. I have just rebased this PR, can you please review it?
Thank you :)

rnetser commented:
@yossisegev kind reminder, pr needs to be marked as `verified`

rnetser commented:
@yossisegev and unresolved comments

yossisegev commented:
@rnetser 
All 19 allegedly unresolved conversations are labeled as outdated, and github doesn't let me reach them and mark them as resolved.
And the change request was also handled (I changed the code the way you requested), but it still appears as an open change request.
Is there anything that can be done about these blockers (none of the solutions I found on google helped)?
Thanks

rnetser commented:
@yossisegev they can be resolved from the PR main gage; I resolved them now.

yossisegev commented:
> @yossisegev they can be resolved from the PR main gage; I resolved them now.

And thank you for doing so, @rnetser :pray: .
I couldn't find a way to resolve them, so it's probably either a matter of permissions, or me trying to do it the wrong way. Anyway - thanks.
